core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/filter_store.go

```diff
@@ -0,0 +1,171 @@
+package logprovider
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/time/rate"
+)
+
+type UpkeepFilterStore interface {
+	InitializeActiveUpkeeps(filters ...upkeepFilter)
+	GetIDs(selector func(upkeepFilter) bool) []*big.Int
+	UpdateFilters(updater func(upkeepFilter, upkeepFilter) upkeepFilter, filters ...upkeepFilter)
+	Has(id *big.Int) bool
+	RangeFiltersByIDs(iterator func(int, upkeepFilter), ids ...*big.Int)
+	GetFilters(selector func(upkeepFilter) bool) []upkeepFilter
+	AddActiveUpkeeps(filters ...upkeepFilter)
+	RemoveActiveUpkeeps(filters ...upkeepFilter)
+	Size() int
+}
+
+var _ UpkeepFilterStore = &upkeepFilterStore{}
+
+type upkeepFilter struct {
+	addr     []byte
+	topics   []common.Hash
+	upkeepID *big.Int
+	// lastPollBlock is the last block number the logs were fetched for this upkeep
+	// used by log event provider.
+	lastPollBlock int64
+	// blockLimiter is used to limit the number of blocks to fetch logs for an upkeep.
+	// used by log event provider.
+	blockLimiter *rate.Limiter
+	// lastRePollBlock is the last block number the logs were recovered for this upkeep
+	// used by log recoverer.
+	lastRePollBlock int64
+}
+
+type upkeepFilterStore struct {
+	lock    *sync.RWMutex
+	filters map[string]upkeepFilter
+}
+
+func NewUpkeepFilterStore() *upkeepFilterStore {
+	return &upkeepFilterStore{
+		lock:    &sync.RWMutex{},
+		filters: map[string]upkeepFilter{},
```

Comment:
why not make the map here? then you could get rid of the `InitializeActiveUpkeeps` function